src/smt/env.cpp

```diff
@@ -306,4 +309,29 @@ bool Env::isBooleanTermSkolem(const Node& k) const
   return d_boolTermSkolems.find(k) != d_boolTermSkolems.end();
 }
 
+Node Env::getSharableFormula(const Node& n) const
+{
+  Node on = n;
+  // these kinds are never sharable
+  std::unordered_set<Kind, kind::KindHashFunction> ks = {Kind::INST_CONSTANT,
+                                                         Kind::DUMMY_SKOLEM};
+  if (!d_options.base.pluginShareSkolems)
+  {
+    // note we only remove purify skolems if the above option is disabled
+    on = SkolemManager::getOriginalForm(n);
```

Comment:
LGTM, but I have one more question: is there ever a situation in which, even if skolem sharing is enabled, we might want to return the original form?

Reply:
Essentially "converting to original form" means replacing all `(@purify t)` with `t`.   When skolems are shared, then this no longer helps make the formula sharable, since we can share the purification skolem anyways.

You could have a heuristic where purification skolems are eliminated and all other skolems are shared, but I think its a lot less intuitive. I don't consider this worthwhile as an option, moreover my intuition is that converting to original form may not be a good idea in general since it "opens the black box" on these skolems.